### PR TITLE
add missing mutex lock

### PR DESCRIPTION
**Pull request by dirk-thomas**
Fixes ros2/ros2#376.

I know this might conflict with the refactoring @Karsten1987 is about to do (#131). If you haven't started the refactoring yet this could maybe be merged before. Otherwise since this patch is minimal I can easily rebase it afterwards.

### CODE REVIEW

**Comment by dirk-thomas** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/2855/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/359/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/2311/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/2994/)

**Comment by Karsten1987** [comment]
I believe that `has_data` is called periodically. With this lock guard, the call would be blocking and potentially breaks any realtime safe execution. Can we replace this with `std::try_lock`?

http://en.cppreference.com/w/cpp/thread/try_lock

**Comment by dirk-thomas** [comment]
I am not sure I understand: what should the code do if the `try_lock` fails?

**Comment by wjwwood** [comment]
> I am not sure I understand: what should the code do if the try_lock fails?

It could busy wait (either by looping there, or triggering the condition variable that the `rmw_wait` function is using so the predicate gets evaluated again), but that doesn't sound very nice.

**Comment by Karsten1987** [comment]
I may not have the full picture here, but my thinking was that if this mutex can't be acquired because the data has yet not been completely read from the DDS, then I would say `has_data()` should return false, indicating that there is no data to be taken.

File: rmw_fastrtps_cpp/src/functions.cpp [hunk 1]
```diff
@@ -648,6 +648,7 @@ class ClientListener : public SubscriberListener
 
   bool hasData()
   {
+    std::lock_guard<std::mutex> lock(internalMutex_);
     return !list.empty();
```
Comment (wjwwood) [inline rmw_fastrtps_cpp/src/functions.cpp]: You could store this state in an atomic boolean variable just before any calls to notify `conditionVariable_` after `list` is modified. That way you would not need a lock here.

Comment (Karsten1987) [inline rmw_fastrtps_cpp/src/functions.cpp]: +1 for the atomic variable

**Comment by dirk-thomas** [comment]
> I may not have the full picture here, but my thinking was that if this mutex can't be acquired because the data has yet not been completely read from the DDS, then I would say has_data() should return false, indicating that there is no data to be taken.

The mutex being currently locked does by no means imply that no data is available. It could be that a different part of the code is holding a lock while data is available.

File: rmw_fastrtps_cpp/src/functions.cpp (hunk 1 above)
Comment (dirk-thomas) [inline rmw_fastrtps_cpp/src/functions.cpp]: Updated in 4be8c428f8f61a1f58335ac7c1d0a26e9008ffc6.

**Comment by dirk-thomas** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/2859/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/363/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/2314/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/2999/)

**Comment by dirk-thomas** [comment]
Ready for re-review.

**Comment by dirk-thomas** [comment]
@Karsten1987 I just noticed that we wanted to wait in case this conflicts with your refactoring. I am sorry I missed that. If it generates more trouble for you please feel free to revert this and I will redo the patch after your PR.

**Comment by Karsten1987** [comment]
my idea was to finish my PR and maybe cherry-pick this one commit. Would
that work or is that no good practice?

File: rmw_fastrtps_cpp/src/functions.cpp [hunk 2]
```diff
@@ -606,6 +607,7 @@ class ClientListener : public SubscriberListener
           } else {
             list.push_back(response);
           }
+          list_has_data_.store(true);
```
Comment (dirk-thomas) [inline rmw_fastrtps_cpp/src/functions.cpp]: With the `list_has_data_` flag now being changed outside of the `conditionMutex_` lock this introduces a race condition with the `rmw_wait()` function. After the function locked the mutex it checks `hasData()` of the listeners and then makes a decision if it has to call `wait`. With this change the boolean flag might change at any time which make `rmw_wait()` call `wait` whe it shouldn't.

See ros2/rmw_fastrtps#147 for the proposed fix.

Before this patch the flag was represented by the state of the container and therefore only changed within the scope of the locked condition.

**Pull request merged**